docs/FIRRTLAnnotations.md

```diff
@@ -0,0 +1,717 @@
+   component.
+2) Annotations, with and without references, are copied to the correct ground
+   type in the `LowerTypes` pass.
+
+Annotations can be parsed using the `--annotation-file` command line argument
+to the `firtool` utility.  Alternatively, we provide a non-standard way of
+encoding annotations in the FIRRTL IR textual representation.  We provide this
+non-standard support primarily to make test writing easier.  As an example of
+this, consider the following JSON annotation file:
+
+```json
+[
+  {
+    "target": "~Foo|Foo",
+    "hello": "world"
+  }
+]
+```
+
+This can be equivalently, in CIRCT, expressed as:
+
+```firrtl
+circuit Foo: %[[{"target":"~Foo|Foo","hello":"world"}]]
+  module Foo:
+    skip
+```
+
+During parsing, annotations are "scattered" into the MLIR representation as
+operation or argument attributes.  As an example of this, the above parses into
+the following MLIR representation:
+
+```mlir
+firrtl.circuit "Foo"  {
+  firrtl.module @Foo() attributes {annotations = [{hello = "world"}]} {
+    firrtl.skip
+  }
+}
+```
+
+Targets without references have their targets stripped during scattering since
+target information is redundant once annotation metadata is attached to the IR.
+Targets with references have the reference portion of the target included in
+the attribute.  The `LowerTypes` pass then uses this reference information to
+attach annotation metadata to only the _lowered_ portion of a targeted circuit
+component.
+
+Annotations are expected to be fully removed via custom transforms, conversion
+to other MLIR operations, or dropped.  E.g., the `ModuleInliner` pass removes
+`firrtl.passes.InlineAnnotation` by inlining annotated modules or instances.
+JSON Annotations map to the builtin MLIR attributes. An annotation is
+implemented using a DictionaryAttr, which holds the class, target, any
+annotation specific data.
+
+## Annotations
+
+Annotations here are written in their JSON format. A "reference target"
+indicates that the annotation could target anything object in the hierarchy,
+although there may be further restrictions in the annotation.
+
+### [BlackBoxInlineAnno](https://www.chisel-lang.org/api/firrtl/latest/firrtl/transforms/BlackBoxInlineAnno.html)
+
+| Property   | Type   | Description                            |
+| ---------- | ------ | -------------                          |
+| class      | string | `firrtl.transforms.BlackBoxInlineAnno` |
+| target     | string | An ExtModule name target               |
+| name       | string | A full path to a file                  |
+| text       | string | Literal verilog code.                  |
+
+Specifies the black box source code (`text`) inline. Generates a file with
+the given `name` in the target directory.
+
+Example:
+```json
+{
+  "class": "firrtl.transforms.BlackBoxInlineAnno",
+  "target": "~Foo|Foo",
+  "name": "blackbox-inline.v",
+  "text": "module ExtInline(); endmodule\n"
+}
+```
```

Comment:
Providing more info on an offline conversation @youngar and I had about this:

There a lot of ways to represent this including tables, prose, JSON examples, JSON schema, other schemas, etc.  Any documentation is better than no documentation and I'm in favor of just keeping this as this solid looking table + prose + example. 

Longer term, there is a lot of information here that _could_ be captured in ODS and generated, including the documentation.  E.g., the unpacking logic for this annotation could be derived from an ODS representation (https://github.com/llvm/circt/blob/6e8be3dc0103c087deb6c97fa072e56b4b2e6794/lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp#L212).